pkgs/build-support/flatten-references-graph/src/flatten_references_graph/pipe.py

```diff
@@ -0,0 +1,80 @@
+from toolz import curried as tlz
+from toolz import curry
+
+from . import lib as lib
+from . import subcomponent as subcomponent
+from .popularity_contest import popularity_contest
+from .split_paths import split_paths
+
+from .lib import (
+    # references_graph_to_igraph
+    debug,
+    pick_attrs
+)
+
+funcs = tlz.merge(
+    pick_attrs(
+        [
+            "flatten",
+            "over",
+            "split_every",
+            "limit_layers",
+            "remove_paths",
+            "reverse"
+        ],
+        lib
+    ),
+    pick_attrs(
+        [
+            "subcomponent_in",
+            "subcomponent_out",
+        ],
+        subcomponent
+    ),
+    {
+        "split_paths": split_paths,
+        "popularity_contest": popularity_contest,
+        "map": tlz.map
+    }
+)
+
+
+@curry
+def nth_or_none(index, xs):
+    try:
+        return xs[index]
+    except IndexError:
+        return None
+
+
+def preapply_func(func_call_data):
+    [func_name, *args] = func_call_data
+    debug("func_name", func_name)
+    debug("args", args)
+    debug('func_name in ["over"]', func_name in ["over"])
+
+    # TODO: these could be handled in more generic way by defining, for each
+    # function, which of the args are expected to be functions which need
+    # pre-applying.
+    if func_name == "over":
+        [first_arg, second_arg] = args
+        args = [first_arg, preapply_func(second_arg)]
+
+    elif func_name == "map":
+        args = [preapply_func(args[0])]
+
+    return funcs[func_name](*args)
+
+
+@curry
+def pipe(pipeline, data):
+    debug("pipeline", pipeline)
+    partial_funcs = list(tlz.map(preapply_func, pipeline))
+    debug('partial_funcs', partial_funcs)
+    return tlz.pipe(
+        data,
+        *partial_funcs
+    )
```

Comment:
I know it's pretty, but just to be able to use this syntax we have to implement these combinators, depend on `toolz`, and wrap our functions with `@curry`. I think they are not pulling their weight for a small script like this.

I bet if we were to write our code like below, this PR would actually be shorter:

```
foo = f(foo, a, b)
foo = g(foo, c)
```

Reply:
>  to use this syntax we have to implement these combinators

Which combinators are you referring to?

Also not sure what you are objecting to: the proposed `dsl` for expressing graph\layer processing pipelines as nix values, or how it is implemented in python?

I agree this is not "idiomatic" python, but maybe implementing functional dsl's is not what it was designed for in the first place. Maybe I should have chose a different language, but I invested quite a bit of energy implementing the `slip_paths` algo using `python-igraph` in the first place, and didn't want to throw it all away  (the idea of the dsl came later)

given the task at hand I believe the use of toolz is most appropriate 

sure, outside the dsl implementation we could avoid using toolz, but since it's already part of the project I couldn't stop myself from using it as every time I see code like:

```python
list = []

for x in xs:
  list.appent(x)
```

I die a little :)

Also, I am not a python dev so idiomatic python doesn't come naturally 

If you'd like to refactor the code a bit to make it more idiomatic I would be happy to accept a PR (you could submit it to my fork).